### ISSUE TITLE
initialState sometimes ignored

### ISSUE THREAD

**Issue by erikerikson** [aws-amplify/amplify-ui#4501]
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have tried disabling all browser extensions or using a different browser
- [X] I have tried deleting the node_modules folder and reinstalling my dependencies
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Authenticator

### How is your app built?

Next.js

### What browsers are you seeing the problem on?

Chrome, Firefox

### Which region are you seeing the problem in?

us-west-2

### Please describe your bug.

Under a sequence of actions, if I use `initialState: 'signIn'` in the options for an `Authenticator` component, I reliably first see the "Create Account" tab during login.

### What's the expected behaviour?

I expect for the "Sign In" tab to be active.

### Help us reproduce the bug!

Preview of README from https://github.com/erikerikson/amplify-auth-example

Steps to reproduce:

1. create a Cognito User Pool as roughly described in `lib/stack.ts` (CDK stack)
1. create a user in that pool (i.e. with name, email, and password)
1. open a terminal in this repo
1. replace `$AWS_REGION`, `$AWS_USER_POOL_ID`, and `$AWS_USER_POOL_WEB_CLIENT_ID` with the values for your user pool
1. run `npm ci`
1. run `npm run dev`
1. open [localhost:3000](http://localhost:3000/), you will be redirected to /signup
1. click the `Sign In` tab (credentials exist)
1. enter your email and credentials
1. click the `Sign Out` button
1. note that you are redirected to /signin but the `Create Account` tab and fields are active.  This is incorrect because the `initialState: 'signIn'` parameter was provided (as can be seen in `/pages/signin.jsx`)

Expected: the screen would have the `Sign In` tab active

### Code Snippet

See working instructions for reproduction, UI code, and CDK code at https://github.com/erikerikson/amplify-auth-example

**Comment by hbuchel** [comment]
Hi @erikerikson I'm having some trouble replicating your steps and running your app locally. With an existing user, I am always redirected to /signup and never to /signin. Are you sure your logic for checking for an "existing" user is correct? 

For example when testing your app, I'm not seeing those values saved in localStorage so even after logging in and out a few times, `existingUser()` in your code always returns false.

**Comment by erikerikson** [comment]
TL;DR: if willing, you could bypass my bug by adding a key with a name starting with `CognitoIdentityServiceProvider` and any value to local storage after you sign out as shown below

@hbuchel Thank you for taking a look.  I'm sorry it didn't work right away, I really expected it would.  I didn't mention that I'm running Ubuntu, if that matters.  Sounds like I very much need to do some cross-platform testing because cookies under http://localhost:3000 is a reliable indicator of previous login for me (ignoring clearing of local storage).

I *would* expect that with an existing Cognito user (that has nonetheless never logged in on the current browser or has cleared 
 local storage) that it would go to `/signup` - that redirect is working as expected.  The desired behavior is that for a probably new user they go straight to "Create Account" and those who have signed in before will go straight to "Sign In".  In my testing, it is faster to skip the sign in process and not have to delete the user repeatedly, so that's why I'm repeatedly using an existing Cognito user.  Regardless, existing users will have had to "clear the site data" (or whatever is appropriate on their platform) and it seems appropriate at that point to treat them as new users until they reestablish site presence on the machine.

For me, with a clean localStorage, when I click "sign up" I am presented with the "Create Account" and sign in with my existing Cognito account, then click "sign out", and then click "sign in" again; `existingUser` is giving true when I log it and I  am properly redirected to `/signin` but am still displayed the "Create Account" tab.

The incorrect result from `existingUser` for you is definitely my bug but you could bypass that bug and hopefully reproduce the issue by adding a key to local storage that starts with "CognitoIdentityServiceProvider" after you click sign out.

![image](https://github.com/aws-amplify/amplify-ui/assets/571200/d9140dce-6985-487a-babe-643878cbbf77)

Reproduction with network requests showing /signin path but display of "Create Account" tab:

![image](https://github.com/aws-amplify/amplify-ui/assets/571200/13ceaefd-7b01-4347-a956-c0143d71739e)

And logging of the result of `existingUser`:

![image](https://github.com/aws-amplify/amplify-ui/assets/571200/32c72baf-82c3-4c38-9758-879a257088fb)

**Comment by erikerikson** [comment]
One additional detail: if, when I have followed these steps and I see the "Create Account" tab displayed on `/signin`, I then hit refresh, I am correctly displayed the "Sign In" tab.  This seems to confirm that there is temporary UI state causing the issue but I really don't understand what is going on and thus I've opened an issue rather than a PR.  :D

**Comment by erikerikson** [comment]
Heather - thank you for overlooking the fact that the use of the environment variables in auth.jsx was clumsy and ineffective.  I am new to front end work and had hard coded my values locally, removed them out of paranoia before pushing to the public repo, and didn't test out that part of the code before sharing.  I've pushed a change that adds this step to the reproduction steps (I will edit the steps above too):

1. replace `$AWS_REGION`, `$AWS_USER_POOL_ID`, and `$AWS_USER_POOL_WEB_CLIENT_ID` with the values for your user pool

I have additionally pushed a change that uses StorageHelper from `@aws-amplify/core`, both passing it in as Storage and then using it to implement existingUser.  The implication seems to be that you had cookies turned off for your browser, sorry for this clumsiness too.  Please pull the reproduction repo and attempt again,

**Comment by reesscot** [comment]
Hi @erikerikson,

The `initialState` is only evaluated on first render of the Authenticator. If you try to change this prop in a subsequent render it will be ignored. However, I think what you are looking for is [state triggers](https://ui.docs.amplify.aws/react/connected-components/authenticator/advanced#full-api), which you can call to transition to the appropriate screen. Try the following:

```jsx
export const authPath = (path, from) => `/${path}?from=${
  encodeURIComponent(from)
}`

export default function Authenticated({ children }) {
  const { authStatus, toSignIn, toSignUp } = useAuthenticator((context) => [context.user])

  const router = useRouter()
  const { pathname } = router
  if (authStatus !== 'authenticated' && authStatus !== 'configuring') {
    if (existingUser()) {
      toSignIn()
      router.replace(authPath('signin', pathname))
    } else {
      toSignUp()
      router.replace(authPath('signup', pathname))
    } 
  } else {
    return children
  }
}
```

That being said, the localstorage keys you are looking for are deleted for me when I sign out, so to test this I had to force `existingUser` to return true and false.

Please let us know if this helps solve your issue.

**Comment by erikerikson** [comment]
Thank you Scott.  Yes, that fixed my issue.

I had expected that separate Authenticator elements on different pages would be distinct objects/elements/controls and thus have separate configuration rather than both be portals onto a shared state machine.  I apologize that I took up your and Heather's time with my misunderstanding.  FWIW, I had originally attempted using those functions (`toSign[In/Up]`) but had expected them to raise the authentication UI (my poor interpretation of "Transitions to _____") and moved on when it didn't work as I hoped.  Perhaps some clarification on that page could guide newbies like me but I only mention in case insight into my misunderstandings might save you and others some time.

Thank you very much for the code, making this change gave me the behavior that I expected, assuming the cookies are only partially cleared during sign out.  Any chance you know the basis for that behavior?
